### PR TITLE
Move to Shared for the Microsoft.Data.SqlClient Reference project

### PR DESCRIPTION
**Pull request by lcheunglci**
This is related to #1261 but it's not in the list.  @cheenamalhotra mentioned that we would like to merge the two reference projects into one as the long term goal is to have one codebase.  I merged the netfx csproj into netcore and combined the Microsoft.Data.SqlClient.cs.  I ran MSBUILD on release and generated the nuget package and compared it before it was merged.  The file size remained the same and I used [NuGet Package Explorer](https://github.com/NuGetPackageExplorer/NuGetPackageExplorer) to view the contents of the package and it looked the same.

### CODE REVIEW

**Comment by Wraith2** [comment]
The implication of this merge would be that all builds of SqlClient have feature parity, and I didn't think that was the case.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs [hunk 1]
```diff
@@ -0,0 +1,2606 @@
...
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlCommandColumnEncryptionSetting.xml' path='docs/members[@name="SqlCommandColumnEncryptionSetting"]/Disabled/*'/>
+        Disabled = 3,
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlCommandColumnEncryptionSetting.xml' path='docs/members[@name="SqlCommandColumnEncryptionSetting"]/Enabled/*'/>
+        Enabled = 1,
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlCommandColumnEncryptionSetting.xml' path='docs/members[@name="SqlCommandColumnEncryptionSetting"]/ResultSetOnly/*'/>
+        ResultSetOnly = 2,
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlCommandColumnEncryptionSetting.xml' path='docs/members[@name="SqlCommandColumnEncryptionSetting"]/UseConnectionSetting/*'/>
+        UseConnectionSetting = 0,
+    }
+#endif
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:623]: Can you merge it with Line 857? It exists for .NET Framework, just not exactly here.
You can change the placement, when merging.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs [hunk 2]
```diff
@@ -0,0 +1,2606 @@
...
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/ColumnEncryptionTrustedMasterKeyPaths/*'/>
+        public static System.Collections.Generic.IDictionary<string, System.Collections.Generic.IList<string>> ColumnEncryptionTrustedMasterKeyPaths { get { throw null; } }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/RegisterColumnEncryptionKeyStoreProviders/*'/>
+        public static void RegisterColumnEncryptionKeyStoreProviders(System.Collections.Generic.IDictionary<string, Microsoft.Data.SqlClient.SqlColumnEncryptionKeyStoreProvider> customProviders) { }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/RegisterColumnEncryptionKeyStoreProvidersOnConnection/*' />
+        public void RegisterColumnEncryptionKeyStoreProvidersOnConnection(System.Collections.Generic.IDictionary<string, Microsoft.Data.SqlClient.SqlColumnEncryptionKeyStoreProvider> customProviders) { }
+#endif // NETCOREAPP
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:892]: Came for this section.
Merge with Line 914.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs [hunk 3]
```diff
@@ -0,0 +1,2606 @@
...
+        protected override System.Data.Common.DbConnection DbConnection { get { throw null; } }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlTransaction.xml' path='docs/members[@name="SqlTransaction"]/IsolationLevel/*'/>
+        public override System.Data.IsolationLevel IsolationLevel { get { throw null; } }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlTransaction.xml' path='docs/members[@name="SqlTransaction"]/Commit/*'/>
+        public override void Commit() { }
+#if NETFRAMEWORK
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlTransaction.xml' path='docs/members[@name="SqlTransaction"]/Dispose/*'/>
+#else
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlTransaction.xml' path='docs/members[@name="SqlTransaction"]/DisposeDisposing/*'/>
+#endif // NETFRAMEWORK
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:1970]: Merge documentation instead.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs [hunk 4]
```diff
@@ -0,0 +1,2606 @@
...
+        public override long GetBytes(int i, long dataIndex, byte[] buffer, int bufferIndex, int length) { throw null; }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlDataReader.xml' path='docs/members[@name="SqlDataReader"]/GetChar/*'/>
+        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
+        public override char GetChar(int i) { throw null; }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlDataReader.xml' path='docs/members[@name="SqlDataReader"]/GetChars/*'/>
+        public override long GetChars(int i, long dataIndex, char[] buffer, int bufferIndex, int length) { throw null; }
+#if NETCOREAPP
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlDataReader.xml' path='docs/members[@name="SqlDataReader"]/GetColumnSchema/*'/>
+        public System.Collections.ObjectModel.ReadOnlyCollection<System.Data.Common.DbColumn> GetColumnSchema() { throw null; }
+#endif // NETCOREAPP
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:1395]: This applies to NETStandard too as we recently enabled, can you check again?

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs [hunk 5]
```diff
@@ -0,0 +1,2606 @@
...
+#endif // NETFRAMEWORK
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlDataAdapter.xml' path='docs/members[@name="SqlDataAdapter"]/OnRowUpdated/*'/>
+        protected override void OnRowUpdated(System.Data.Common.RowUpdatedEventArgs value) { }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlDataAdapter.xml' path='docs/members[@name="SqlDataAdapter"]/OnRowUpdating/*'/>
+        protected override void OnRowUpdating(System.Data.Common.RowUpdatingEventArgs value) { }
+        object System.ICloneable.Clone() { throw null; }
+#if NETFRAMEWORK
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlDataAdapter.xml' path='docs/members[@name="SqlDataAdapter"]/TerminateBatching/*'/>
+        protected override void TerminateBatching() { }
+#endif // NETFRAMEWORK
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:1353]: Merge the 2 if-defs sections in one - add to end of class.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs [hunk 6]
```diff
@@ -0,0 +1,2606 @@
...
+        [System.ComponentModel.RefreshPropertiesAttribute(System.ComponentModel.RefreshProperties.All)]
+        public bool PersistSecurityInfo { get { throw null; } set { } }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnectionStringBuilder.xml' path='docs/members[@name="SqlConnectionStringBuilder"]/PoolBlockingPeriod/*'/>
+        [System.ComponentModel.DisplayNameAttribute("Pool Blocking Period")]
+        [System.ComponentModel.RefreshPropertiesAttribute(System.ComponentModel.RefreshProperties.All)]
+#if NETFRAMEWORK
+        public Microsoft.Data.SqlClient.PoolBlockingPeriod PoolBlockingPeriod { get { throw null; } set { } }
+#else
+        public PoolBlockingPeriod PoolBlockingPeriod { get { throw null; } set { } }
+#endif // NETFRAMEWORK
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:1231]: Doesn't need if-defs, use any one.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs [hunk 7]
```diff
@@ -0,0 +1,2606 @@
...
+#if NETCOREAPP || NETSTANDARD21_AND_ABOVE
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnectionStringBuilder.xml' path='docs/members[@name="SqlConnectionStringBuilder"]/AttestationProtocol/*' />
+        [System.ComponentModel.DisplayNameAttribute("Attestation Protocol")]
+        [System.ComponentModel.RefreshPropertiesAttribute(System.ComponentModel.RefreshProperties.All)]
+        public Microsoft.Data.SqlClient.SqlConnectionAttestationProtocol AttestationProtocol { get { throw null; } set { } }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnectionStringBuilder.xml' path='docs/members[@name="SqlConnectionStringBuilder"]/EnclaveAttestationUrl/*'/>
+        [System.ComponentModel.DisplayNameAttribute("Enclave Attestation Url")]
+        [System.ComponentModel.RefreshPropertiesAttribute(System.ComponentModel.RefreshProperties.All)]
+        public string EnclaveAttestationUrl { get { throw null; } set { } }
+#endif
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:1154]: Should also apply to NETFRAMEWORK, please check again.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs [hunk 8]
```diff
@@ -0,0 +1,2606 @@
...
+        public void Open(SqlConnectionOverrides overrides) { }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/OpenAsync/*'/>
+        public override System.Threading.Tasks.Task OpenAsync(System.Threading.CancellationToken cancellationToken) { throw null; }
+#if NETFRAMEWORK
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/RegisterColumnEncryptionKeyStoreProviders/*'/>
+        public static void RegisterColumnEncryptionKeyStoreProviders(System.Collections.Generic.IDictionary<string, Microsoft.Data.SqlClient.SqlColumnEncryptionKeyStoreProvider> customProviders) { }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/RegisterColumnEncryptionKeyStoreProvidersOnConnection/*' />
+        public void RegisterColumnEncryptionKeyStoreProvidersOnConnection(System.Collections.Generic.IDictionary<string, Microsoft.Data.SqlClient.SqlColumnEncryptionKeyStoreProvider> customProviders) { }
+#endif // NETFRAMEWORK
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:1051]: It applies to all supported, please check again?

**Review by cheenamalhotra — COMMENTED** [review]
Please organize merges better in classes, you can search for method/type by their name before categorizing them and put them in 1 location only.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs (hunk 8 above)
Comment (lcheunglci) [reply src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:1051]: yes, they are in both. updated in the next commit.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs [hunk 9]
```diff
@@ -0,0 +1,2569 @@
...
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlClientLogger.xml' path='docs/members[@name="SqlClientLogger"]/IsLoggingEnabled/*'/>
+        public bool IsLoggingEnabled { get { throw null; } }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlClientLogger.xml' path='docs/members[@name="SqlClientLogger"]/LogAssert/*'/>
+        public bool LogAssert(bool value, string type, string method, string message) { throw null; }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlClientLogger.xml' path='docs/members[@name="SqlClientLogger"]/LogError/*'/>
+        public void LogError(string type, string method, string message) { }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlClientLogger.xml' path='docs/members[@name="SqlClientLogger"]/LogInfo/*'/>
+        public void LogInfo(string type, string method, string message) { }
+    }
+#endif //NETFRAMEWORK
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:453]: SqlClientLogger exists in both projects.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs [hunk 10]
```diff
@@ -0,0 +1,2569 @@
...
+        public override System.Data.Common.DbDataSourceEnumerator CreateDataSourceEnumerator() { throw null; }
+
+#endif // NETFRAMEWORK
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlClientFactory.xml' path='docs/members[@name="SqlClientFactory"]/CreateParameter/*'/>
+        public override System.Data.Common.DbParameter CreateParameter() { throw null; }
+#if NETFRAMEWORK
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlClientFactory.xml' path='docs/members[@name="SqlClientFactory"]/CreatePermission/*'/>
+        public override System.Security.CodeAccessPermission CreatePermission(System.Security.Permissions.PermissionState state) { throw null; }
+        object System.IServiceProvider.GetService(System.Type serviceType) { throw null; }
+#endif // NETFRAMEWORK
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:436]: Please collect all if-defs together in 1 block.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs [hunk 11]
```diff
@@ -0,0 +1,2569 @@
...
+        [System.ComponentModel.DefaultValueAttribute(null)]
+        public static System.Collections.Generic.IDictionary<string, System.Collections.Generic.IList<string>> ColumnEncryptionTrustedMasterKeyPaths { get { throw null; } }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/RegisterColumnEncryptionKeyStoreProviders/*'/>
+        public static void RegisterColumnEncryptionKeyStoreProviders(System.Collections.Generic.IDictionary<string, Microsoft.Data.SqlClient.SqlColumnEncryptionKeyStoreProvider> customProviders) { }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/RegisterColumnEncryptionKeyStoreProvidersOnConnection/*' />
+        public void RegisterColumnEncryptionKeyStoreProvidersOnConnection(System.Collections.Generic.IDictionary<string, Microsoft.Data.SqlClient.SqlColumnEncryptionKeyStoreProvider> customProviders) { }
+#endif // NETCOREAPP || NETFRAMEWORK
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:882]: Please check this block thoroughly before PR, these apply to NETSTANDARD too.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs [hunk 12]
```diff
@@ -0,0 +1,2569 @@
...
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/BeginTransactionIsoTransactionName/*'/>
+        public Microsoft.Data.SqlClient.SqlTransaction BeginTransaction(System.Data.IsolationLevel iso, string transactionName) { throw null; }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/BeginTransactionTransactionName/*'/>
+        public Microsoft.Data.SqlClient.SqlTransaction BeginTransaction(string transactionName) { throw null; }
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnection.xml' path='docs/members[@name="SqlConnection"]/ChangeDatabase/*'/>
+        public override void ChangeDatabase(string database)
+        {
+#if NETCOREAPP
+            throw null;
+#endif
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:964]: We can throw null in both so this if-def is not needed.
Similarly in other places.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs [hunk 13]
```diff
@@ -0,0 +1,2569 @@
...
+        [System.ComponentModel.DisplayNameAttribute("Connect Timeout")]
+        [System.ComponentModel.RefreshPropertiesAttribute(System.ComponentModel.RefreshProperties.All)]
+        public int ConnectTimeout { get { throw null; } set { } }
+#if NETFRAMEWORK
+        /// <include file='../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnectionStringBuilder.xml' path='docs/members[@name="SqlConnectionStringBuilder"]/ContextConnection/*'/>
+        [System.ComponentModel.DisplayNameAttribute("Context Connection")]
+        [System.ComponentModel.RefreshPropertiesAttribute(System.ComponentModel.RefreshProperties.All)]
+        [System.ObsoleteAttribute("ContextConnection has been deprecated.  SqlConnection will ignore the 'Context Connection' keyword.")]
+        public bool ContextConnection { get { throw null; } set { } }
+#endif // NETFRAMEWORK
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:1109]: All `#if NETFRAMEWORK` blocks in this type can be placed together and combined into 1 block:
- ConnectionReset
- ContextConnection
- NetworkLibrary
- TransparentNetworkIPResolution

**Review by cheenamalhotra — COMMENTED** [review]
Architecture looks good, few more changes needed.

File: src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs (hunk 9 above)
Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:453]: I think the code exists in NetCore too, it was not exposed for NetCore/NetStandard 🤔 

@JRahnama @DavoudEshtehari it probably got missed with EventSource changes, any concerns exposing this for NetCore/NetStandard refs?

Comment (lcheunglci) [reply src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:448]: I made all the changes except for this one. Once I get a confirmation that I can expose this logger, I'll make another commit to remove the ifdef.

Comment (lcheunglci) [reply src/Microsoft.Data.SqlClient/ref/Microsoft.Data.SqlClient.cs:448]: I was going to open another PR to merge the `SqlClientLogger.cs`, but it's already in shared src and both netcore and netfx are using it and I see references in netcore, so I think I won't need to open another PR and can make the changes here in the next commit to remove the `ifdef`

**Comment by lcheunglci** [comment]
I did notice there's already a `Microsoft.Data.SqlClient.NetStandard.cs` and `Microsoft.Data.SqlClient.Manual.cs`. I think having one for net framework in `Microsoft.Data.SqlClient.NetFx.cs` would make sense for clarity. I'll split it up in the next commit.

**Comment by JRahnama** [comment]
Closing this PR as it not in a mergeable state.

**Pull request closed without merging**